js/ai/src/formats/enum.ts

```diff
@@ -0,0 +1,40 @@
+/**
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { GenkitError } from '@genkit-ai/core';
+import type { Formatter } from './types';
+
+export const enumParser: Formatter = (request) => {
+  const schemaType = request.output?.schema?.type;
+  if (schemaType && schemaType !== 'string' && schemaType !== 'enum') {
+    throw new GenkitError({
+      status: 'INVALID_ARGUMENT',
+      message: `Must supply a 'string' or 'enum' schema type when using the enum parser format.`,
+    });
+  }
+
+  let instructions: boolean | string = false;
+  if (request.output?.schema?.enum) {
+    instructions = `Output should be ONLY one of the following enum values. Do not output any additional information or add quotes.\n\n${request.output?.schema?.enum.map((v) => v.toString()).join('\n')}`;
```

Comment:
Have you given any thought to how we might integrate this with the plan to allow devs to specify their own schema instruction message?

Reply:
Yes, the idea is that the format provides a "base layer" of instructions that are set to be augmented into the prompt; however, the `instructions` field will be added to `output` and you can either override it with your own custom instructions or set it to `false` to opt-out of automatic instructions altogether if you included it otherwise in your prompt. For example:

```ts
ai.generate({
  prompt: "Output some JSON in the format `{foo: 'bar'}`."
  output: {format: 'json', instructions: false}
});
```

Reply:
Do we have a placeholder to inject the schema into the prompt along with the custom instructions? Sorry if I missed it.

Reply:
No placeholder (since if you just wanted to insert the json schema raw you'd probably do that). For custom formatters, they have access to the whole request when returning their instructions. For users, I'm thinking we probably allow something like:

```ts
ai.generate({
  format: {
    schema: MySchema,
    instructions: (req) => `Use this format:\n\n${formatSchemaForInstructions(req.output.schema)}`,
  }
});
```